domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java

```diff
@@ -1097,26 +1097,51 @@ public static void setOptimizedSqlSize(int size) {
     }
   }
 
-  private static <T, C> C collection(Class<T> clazz, Class<C> collectionClazz, ResultSet resultSet, int limit)
+  private static <T, C> C collection(Class<T> clazz, Class<C> collectionClazz, ResultSet resultSet,
+    int offset, int limit)
       throws ReflectiveOperationException, IOException {
 
     List<JsonObject> jsonList = resultSet.getRows();
-    List<T> recordList = new ArrayList<>(jsonList.size());
+    int resultSize = jsonList.size();
+    List<T> recordList = new ArrayList<>(resultSize);
     int totalRecords = 0;
     for (JsonObject object : jsonList) {
       String jsonb = object.getString(JSON_COLUMN);
       recordList.add(OBJECT_MAPPER.readValue(jsonb, clazz));
       totalRecords = object.getInteger("count");
     }
 
-    // full table scan was stopped without total records calculation.
-    if (totalRecords == 0 && jsonList.size() == limit) {
-      totalRecords = 999999999;  // unknown total
-    }
+    totalRecords = getTotalRecords(resultSize, totalRecords, offset, limit);
 
     return collection(collectionClazz, recordList, totalRecords);
   }
 
+  /**
+   * Function to correct estimated result count:
+   * If the resultsCount is equal to 0, the result should be not more than offset
+   * If the resultsCount is equal to limit, the result should be not less than offset
+   * Otherwise it should be equal to offset + resultsCount
+   *
+   * @param resultsCount the count of rows, that are returned from database
+   * @param estimateCount the estimate result count from returned by database
+   * @param offset database offset
+   * @param limit database limit
+   * @return corrected results count
+   */
+  static int getTotalRecords(int resultsCount, int estimateCount, int offset, int limit) {
+    if (limit == 0) {
+      return estimateCount;
+    }
+    if (resultsCount == 0) {
+      estimateCount = Math.min(offset, estimateCount);
+    } else if (resultsCount == limit) {
+      estimateCount = Math.max(offset + limit, estimateCount);
+    } else {
+      estimateCount = offset + resultsCount;
+    }
+    return estimateCount;
```

Comment:
There is no need to change `estimateCount`, you can directly return the result, for example

    return Math.min(offset, estimateCount);

This also avoids parameter assignment: https://checkstyle.org/config_coding.html#ParameterAssignment